[WK2] Allow moving objects into ArgumentCoder encoding functions

zdobersek opened the pull request:
#### afff11b97280d9bb26274dfad5ca123f8bffffae
<pre>
[WK2] Allow moving objects into ArgumentCoder encoding functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=247525">https://bugs.webkit.org/show_bug.cgi?id=247525</a>

Reviewed by Alex Christensen and Kimmo Kinnunen.

Adjust current encode() functions for different ArgumentCoder
specializations so that rvalue references to a given object can be
passed in, enabling different resources to be more optimally handled.
This mostly concerns containers and other wrapper types like
std::optional&lt;&gt; and std::variant&lt;&gt;. In changed encode() functions,
universal references are used to accept the passed-in value, and
std::forward_like&lt;&gt; is used to apply the value category from the
universal reference to the contained object that&apos;s passed forward into
the encoding process.

std::forward_like&lt;&gt; is a C++23 feature, so not in line yet with the
current C++ standard in use in the project. A fallback implementation is
added in StdLibExtras.h, properly guarded with the relevant feature-test
macro.

As an example, this can simplify encoding of Unix file descriptors or
containers of such items. When such objects can be consumed through
these rvalue references, the underlying resources can completely avoid
any otherwise-necessary duplication.

Testing facilities are added, covering different types that now support
encoding of objects passed through rvalue references. For each type,
testing is done for lvalues, rvalues and moved-in rvalues of that type.
HashTable-based types are skipped for now because of issues in
constructing appropriate values of those types.

* Source/WTF/wtf/StdLibExtras.h:
(std::forward_like):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::ArgumentCoder&lt;std::optional&lt;T&gt;&gt;::encode):
(IPC::ArgumentCoder&lt;std::optional&lt;T&gt;&gt;::decode):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp: Added.
(TestWebKitAPI::EncodingCounter::CounterValues::CounterValues):
(TestWebKitAPI::EncodingCounter::CounterValues::operator== const):
(TestWebKitAPI::EncodingCounter::EncodingCounter):
(TestWebKitAPI::EncodingCounter::encode const):
(TestWebKitAPI::EncodingCounter::encode):
(TestWebKitAPI::EncodingCounter::decode):
(TestWebKitAPI::PrintTo):
(TestWebKitAPI::ArgumentCoderEncodingCounterTest::ArgumentCoderEncodingCounterTest):
(TestWebKitAPI::ArgumentCoderEncodingCounterTest::testEncoding):
(TestWebKitAPI::TEST_P):

Canonical link: <a href="https://commits.webkit.org/257073@main">https://commits.webkit.org/257073@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/20c9ac55c7b301dbddfd79682263d192a9bd47ec

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/97411 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6670 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/30575 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106930 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/167194 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/101376 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6993 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/35617 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89819 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103603 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/103072 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/5250 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/84043 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/32251 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/87116 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88918 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/75176 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/88325 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/683 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/20384 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/83981 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/668 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21852 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28484 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4860 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/5475 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/44331 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/86807 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1913 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/41205 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19506 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -491,6 +516,14 @@ template<typename KeyArg, typename MappedArg, typename HashArg, typename KeyTrai
+    static void encode(Encoder& encoder, HashMapType&& hashMap)
+    {
+        encoder << static_cast<uint32_t>(hashMap.size());
+        for (typename HashMapType::iterator it = hashMap.begin(), end = hashMap.end(); it != end; ++it)
```
Can we use c++11-style iteration instead?

zdobersek replied:
Yes, of course.

achristensen07 commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -674,26 +715,27 @@ template<typename ErrorType> struct ArgumentCoder<Expected<void, ErrorType>> {
-    static void encode(Encoder& encoder, const std::variant<Types...>& variant)
+    template<typename Encoder, typename T>
+    static auto encode(Encoder& encoder, T&& variant)
+        -> std::enable_if_t<std::is_same_v<std::remove_cvref_t<T>, std::variant<Types...>>>
```
An idiom we use elsewhere is to add a template parameter of std::enable_if_t<something>* = nullptr instead of using it as a return type.  It makes it more straightforward to see what the return type is.

zdobersek replied:
OK, I think that should be doable everywhere.

achristensen07 commented on Source/WebKit/Platform/IPC/ArgumentCoders.h:
```diff
@@ -575,6 +608,14 @@ template<typename KeyArg, typename HashArg, typename KeyTraitsArg> struct Argume
     }
 
+    template<typename Encoder>
+    static void encode(Encoder& encoder, HashCountedSetType&& hashCountedSet)
```
It would be nice if we could use consistent styling.  Now sometimes we have 2 encoders, and sometimes we have an encoder that takes a universal reference.

zdobersek replied:
It depends on how the move is achievable. For instance, with `std::pair` the single `encode()` that utilizes the forwarding reference works fine because of the `std::get<>` overloads. `WTF::KeyValuePair` doesn't have such luck, so a separate `encode()` overload is used to move out the key and the value.

It's probably possible to get this working even for things like `WTF::KeyValuePair`, i.e. translate the value category of the forwarding reference into an appropriate `std::forward<>` call on the contained objects, I just haven't researched it yet.

zdobersek replied:
When moving out the encoded-object's members, the forwarding reference's value category can be applied simply through `std::forward<T>(object).key`. But this doesn't work for containers like vectors or hash maps.

What's necessary there is `std::forward_like`. It's only available in C++23, but we can carry an implementation in `StdLibExtras.h`.
https://en.cppreference.com/w/cpp/utility/forward_like

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
```diff
@@ -0,0 +1,153 @@
+    CounterValues& m_counterValues;
+};
+
+TEST(ArgumentCoder, MoveSemantics)
```
This could be
```
class ArgumentCoderTests : public testing::Test {
protected:
   void runTest(...) ...
}

TEST_F(ArgumentCoderTest, EncodeRValueOptional) {
 ....
}
...
```

This way if the test ever fails we know what fails immediately.
Otherwise it's a major debugging task that involves a backtrace to understand which assert fails.

(E.g. we write tests because we expect them to fail, and if a test can fail, then the failure should be understandable.)

zdobersek replied:
Moved the tests into a testing suite.

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
```diff
@@ -0,0 +1,153 @@
+                auto object = createFunctor(counterValues);
+                encoder << object;
+
+                counterValues.assertEqual(expectedEncodingCount, 0);
```
in similar vein, if possible:
```
ASSERT_EQ(counterValues, EncodingCounter { expectedEncodingCount, 0 });
```

zdobersek replied:
Possible.

kkinnunen-apple commented on Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp:
```diff
@@ -0,0 +1,153 @@
+        });
+}
+
+} // namespace TestWebKitAPI
```
Something that might be useful would be to try if encoding containers of const types still compiles. This isn't invoked much in current webkit, and I'm not 100% sure decoding works, but just if you're looking more to cover.

zdobersek replied:
You mean something like `Vector<const T>`? My hunch is reference collapsing would yield a const lvalue reference. I can add tests.

kkinnunen-apple reviewed (approved):
Good by me. @achristensen07 should take a look, GitHub needs the ok.

zdobersek committed:
[WK2] Allow moving objects into ArgumentCoder encoding functions
https://bugs.webkit.org/show_bug.cgi?id=247525

Reviewed by Alex Christensen and Kimmo Kinnunen.

Adjust current encode() functions for different ArgumentCoder
specializations so that rvalue references to a given object can be
passed in, enabling different resources to be more optimally handled.
This mostly concerns containers and other wrapper types like
std::optional<> and std::variant<>. In changed encode() functions,
universal references are used to accept the passed-in value, and
std::forward_like<> is used to apply the value category from the
universal reference to the contained object that's passed forward into
the encoding process.

std::forward_like<> is a C++23 feature, so not in line yet with the
current C++ standard in use in the project. A fallback implementation is
added in StdLibExtras.h, properly guarded with the relevant feature-test
macro.

As an example, this can simplify encoding of Unix file descriptors or
containers of such items. When such objects can be consumed through
these rvalue references, the underlying resources can completely avoid
any otherwise-necessary duplication.

Testing facilities are added, covering different types that now support
encoding of objects passed through rvalue references. For each type,
testing is done for lvalues, rvalues and moved-in rvalues of that type.
HashTable-based types are skipped for now because of issues in
constructing appropriate values of those types.

* Source/WTF/wtf/StdLibExtras.h:
(std::forward_like):
* Source/WebKit/Platform/IPC/ArgumentCoders.h:
(IPC::ArgumentCoder<std::optional<T>>::encode):
(IPC::ArgumentCoder<std::optional<T>>::decode):
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/IPC/ArgumentCoderTests.cpp: Added.
(TestWebKitAPI::EncodingCounter::CounterValues::CounterValues):
(TestWebKitAPI::EncodingCounter::CounterValues::operator== const):
(TestWebKitAPI::EncodingCounter::EncodingCounter):
(TestWebKitAPI::EncodingCounter::encode const):
(TestWebKitAPI::EncodingCounter::encode):
(TestWebKitAPI::EncodingCounter::decode):
(TestWebKitAPI::PrintTo):
(TestWebKitAPI::ArgumentCoderEncodingCounterTest::ArgumentCoderEncodingCounterTest):
(TestWebKitAPI::ArgumentCoderEncodingCounterTest::testEncoding):
(TestWebKitAPI::TEST_P):

Canonical link: https://commits.webkit.org/257073@main
